Shorten #file and add #filePath (behind an experimental flag)

beccadax opened the pull request:
`#file` includes an entire gigantic path whenever it is used, and it is often used implicitly. This has two negative effects: It wastes space on something that is almost never needed and it invisibly embeds information in the binary, such as the developer's username, that they may want to keep private.

This PR changes the behavior of `#file` so that it evaluates to a short string containing the module name and filename, but not the full path. It also adds `#filePath`, which continues to give you the absolute path.

These changes are hidden behind a feature flag, `-enable-experimental-concise-pound-file`, because making them permanent [will require an evolution proposal](https://forums.swift.org/t/pitch-concise-magic-file-names/31297).

This change includes some basic tests with the flag enabled, but they're not comprehensive; for instance, code completion isn't tested anywhere. The implementation for the proposal should update existing `#file` tests and add matching `#filePath` tests to ensure adequate coverage.

Progresses towards fixing rdar://problem/56816166.

beccadax committed:
Make #file only include the filename

This change may have significant code size benefits.

brentdax commented:
@swift-ci please benchmark

brentdax commented:
Okay, "up to 6% code size" is probably a fair summary of that. Let's try compiler performance and look at number of bytes output as a rough proxy for code size of the source compatibility suite.

@swift-ci please test compiler performance

beccadax committed:
Include module name in string

Along with a not-yet-usable way to get back to the old behavior.

beccadax committed:
Control #file behavior with a command line option

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please check compiler performance

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
macOS test failures are examples of the feature working correctly, but affecting tests I didn't update. I'm not sure why they didn't run locally. 🤷‍♂️

brentdax commented on include/swift/Basic/LangOptions.h:
```diff
@@ -94,6 +94,10 @@ namespace swift {
 
+    /// If true, '#file' evaluates to the full path rather than a
+    /// human-readable string.
+    bool MagicFileIdentifierEvaluatesToPath = false;
```
Without an evolution proposal, this will need to be `true`.

beccadax committed:
Turn new #file behavior off by default

brentdax commented:
@swift-ci please test

beccadax committed:
Flip default in test

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please smoke test

beccadax committed:
Introduce a separate #filePath, remove -pound-file

This makes the path behavior more first-class. The feature is now hidden behind an experimental flag, -enable-experimental-concise-pound-file.

brentdax commented:
@swift-ci please test

brentdax commented:
The test failure is because I need to re-gyb SwiftSyntax files.

brentdax commented:
apple/swift-syntax#185

@swift-ci please test

brentdax commented:
apple/swift-syntax#185

@swift-ci please test

jckarter reviewed (approved):
LGTM.

brentdax commented:
apple/swift-syntax#185

@swift-ci please clean test
